src/statement.d

```diff
@@ -4863,6 +4863,17 @@ public:
                 exp = new CastExp(loc, exp, t);
                 exp = exp.semantic(sc);
             }
+
+            /+ Semantic checking is skipped for the calls to _d_monitorenter/exit below,
+             + however they may write to the __monitor ptr of the object (argument type is mutable Object).
+             + Ensure here that the synchronization object is mutable.
+             +/
+            if (!exp.type.isMutable())
+            {
+                error("can only synchronize on a mutable object, not on '%s' of type '%s'", exp.toChars(), exp.type.toChars());
```

Comment:
Perhaps DMD does not optimize enough for this to result in bad codegen. But for LDC, this will have to be an error, because it results in crashing code and was never supported.